cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -644,28 +643,24 @@ public void getProgramStatusSchedules(HttpRequest request, HttpResponder respond
       : new ApplicationId(triggerNamespaceId, triggerAppName, triggerAppVersion);
     final ProgramId triggerProgramId = appId.program(ProgramType.valueOfCategoryName(triggerProgramType),
                                                      triggerProgramName);
-    Collection<ProgramScheduleRecord> schedules = programScheduler.findSchedules(triggerProgramId.toString());
-    if (triggerProgramStatuses != null) {
-      final Set<co.cask.cdap.api.ProgramStatus> queryProgramStatuses = new HashSet<>();
-      try {
-        for (String status : triggerProgramStatuses.split(",")) {
-          queryProgramStatuses.add(co.cask.cdap.api.ProgramStatus.valueOf(status));
-        }
-      } catch (Exception e) {
-        throw new BadRequestException(String.format("Unable to parse program statuses '%s'. Must be comma separated " +
-                                                      "valid ProgramStatus names such as COMPLETED, FAILED, KILLED.",
-                                                    triggerProgramStatuses), e);
-      }
-      Iterator<ProgramScheduleRecord> iterator = schedules.iterator();
-      while (iterator.hasNext()) {
-        // Remove the schedule if none of its triggering program statuses is in the query program statuses set
-        Set<co.cask.cdap.api.ProgramStatus> triggerProgramStatusesSet =
-          ((ProgramStatusTrigger) iterator.next().getSchedule().getTrigger()).getProgramStatuses();
-        if (Sets.intersection(triggerProgramStatusesSet, queryProgramStatuses).isEmpty()) {
-          iterator.remove();
-        }
+
+    Collection<ProgramScheduleRecord> schedules = new HashSet<>();
+    Set<co.cask.cdap.api.ProgramStatus> queryProgramStatuses = new HashSet<>();
+    if (triggerProgramStatuses == null) {
+      queryProgramStatuses = ImmutableSet.of(co.cask.cdap.api.ProgramStatus.COMPLETED,
+                                             co.cask.cdap.api.ProgramStatus.FAILED,
+                                             co.cask.cdap.api.ProgramStatus.KILLED);
+    } else {
+      for (String status : triggerProgramStatuses.split(",")) {
+        queryProgramStatuses.add(co.cask.cdap.api.ProgramStatus.valueOf(status));
       }
     }
+
+    for (co.cask.cdap.api.ProgramStatus status : queryProgramStatuses) {
+      String triggerKey = Schedulers.triggerKeyForProgramStatus(triggerProgramId, status);
+      schedules.addAll(programScheduler.findSchedules(triggerKey));
```

Comment:
Because of this line, I needed to implement the hashCode, equals, methods for `ProgramScheduleMeta` and `ProgramScheduleRecord`